pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/utils/RegexpMatcher.java

```diff
@@ -0,0 +1,170 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.utils.nativefst.utils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import org.apache.pinot.segment.local.utils.nativefst.FST;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.Automaton;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.CharacterRunAutomaton;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.RegExp;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.State;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.Transition;
+
+
+/**
+ * RegexpMatcher is a helper to retrieve matching values for a given regexp query.
+ * Regexp query is converted into an automaton and we run the matching algorithm on FST.
+ *
+ * Two main functions of this class are
+ *   regexMatchOnFST() Function runs matching on FST (See function comments for more details)
+ *   match(input) Function builds the automaton and matches given input.
+ */
+public class RegexpMatcher {
+  private final String _regexQuery;
+  private final FST _fst;
+  private final Automaton _automaton;
+
+  public RegexpMatcher(String regexQuery, FST fst) {
+    _regexQuery = regexQuery;
+    _fst = fst;
+
+    _automaton = new RegExp(_regexQuery).toAutomaton();
+  }
+
+  public static List<Long> regexMatch(String regexQuery, FST fst) {
```

Comment:
Does this need to produce `List<Long>` - can it avoid boxing and yield an iterator instead?

Reply:
Returning an iterator would still need building a list internally, right?

Reply:
I believe `regexMatchOnFST` could be implemented as an iterator, yielding whenever `endNodes` is added to. But my main concern is the boxing, not the eagerness.

Reply:
Simpler would be to take an `IntConsumer` so it could be directly appended to the resultant bitmap without this depending on RoaringBitmap. So something like:

```
static void regexMatch(String regexQuery, FST fst, IntConsumer dest) {
    ...
   dest.accept(_fst.getOutputSymbol(path._fstArc));
}
...
RoaringBitmapWriter<MutableRoaringBitmap> writer = RoaringBitmapWriter.bufferWriter().get();
regexMatch(regex, fst, writer::add);